lib/rexml/parsers/baseparser.rb

```diff
@@ -198,183 +207,188 @@ def pull_event
+                match[3] = match[3][1..-2] # PUBID
+                match[4] = match[4][1..-2] # HREF
+                match.delete_at(5) if match.size > 5 # Chop out NDATA decl
+                # match is [ :entity, name, PUBLIC, pubid, href(, ndata)? ]
               else
-                message = "#{base_error_message}: invalid declaration name"
+                match[2] = match[2][1..-2]
+                match.pop if match.size == 4
+                # match is [ :entity, name, value ]
               end
-              raise REXML::ParseException.new(message, @source)
-            end
-            name = parse_name(base_error_message)
-            id = parse_id(base_error_message,
-                          accept_external_id: true,
-                          accept_public_id: true)
-            unless @source.match(/\A\s*>/um, true)
-              message = "#{base_error_message}: garbage before end >"
-              raise REXML::ParseException.new(message, @source)
+              match << '%' if ref
+              return match
+            elsif @source.match("ATTLIST", true)
+              md = @source.match(ATTLISTDECL_END, true)
+              raise REXML::ParseException.new( "Bad ATTLIST declaration!", @source ) if md.nil?
+              element = md[1]
+              contents = md[0]
+
+              pairs = {}
+              values = md[0].scan( ATTDEF_RE )
+              values.each do |attdef|
+                unless attdef[3] == "#IMPLIED"
+                  attdef.compact!
+                  val = attdef[3]
+                  val = attdef[4] if val == "#FIXED "
+                  pairs[attdef[0]] = val
+                  if attdef[0] =~ /^xmlns:(.*)/
+                    @nsstack[0] << $1
+                  end
+                end
+              end
+              return [ :attlistdecl, element, pairs, contents ]
+            elsif @source.match("NOTATION", true)
+              base_error_message = "Malformed notation declaration"
+              unless @source.match(/\s+/um, true)
+                if @source.match(">")
+                  message = "#{base_error_message}: name is missing"
+                else
+                  message = "#{base_error_message}: invalid name"
+                end
+                @source.string = " <!NOTATION" + @source.buffer
+                raise REXML::ParseException.new(message, @source)
+              end
+              name = parse_name(base_error_message)
+              id = parse_id(base_error_message,
+                            accept_external_id: true,
+                            accept_public_id: true)
+              unless @source.match(/\s*>/um, true)
+                message = "#{base_error_message}: garbage before end >"
+                raise REXML::ParseException.new(message, @source)
+              end
+              return [:notationdecl, name, *id]
+            elsif md = @source.match( /--(.*?)-->/um, true)
+              case md[1]
+              when /--/, /-\z/
```

Comment:
```suggestion
              if md[1].end_with?("-")
```

Reply:
@kou 
With the above changes, the test code errors.
```
Failure: test_bad_markup(REXMLTests::Tester):
  Parse "<a><!-- -- --></a>" should have failed!.
  <REXML::ParseException> exception was expected but none was thrown.
/Users/naitoh/ghq/github.com/naitoh/rexml/test/test_core.rb:61:in `block in test_bad_markup'
     58: # FIXME      '<a' + [0x0371].pack('U') + ' />',
     59: # FIXME      '<a a' + [0x0371].pack('U') + '="" />',
     60:       ].each do |src|
  => 61:         assert_raise( ParseException, %Q{Parse #{src.inspect} should have failed!} ) do
     62:           Document.new(src)
     63:         end
     64:       end
/Users/naitoh/ghq/github.com/naitoh/rexml/test/test_core.rb:60:in `each'
/Users/naitoh/ghq/github.com/naitoh/rexml/test/test_core.rb:60:in `test_bad_markup'
```

https://github.com/ruby/rexml/blob/fb7ba27594ce15e2a0a566c837355cb4beb4db14/test/test_core.rb#L47-L63


https://www.w3.org/TR/2006/REC-xml11-20060816/#NT-Comment
```
[15]   Comment   ::=   '<!--' ((Char - '-') | ('-' (Char - '-')))* '-->'
```

`<!-- -- -->` is an XML violation and should raise a REXML::ParseException exception.

Reply:
Ah, sorry. I missed the case.